### ISSUE TITLE
Recursive alias reported as not iterable in certain circumstances

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#1299]
Not sure if it counts as a bug, but the error is certainly confusing.
For a minimal example, let's define a recursive alias like this:
```py
from typing import Optional

LinkedList = Optional[tuple[int, "LinkedList"]]
```
Suppose that we want to make a generator function that yields all values in a list:

This typechecks, but is not right because we forgot to reassign `xs`.
```py
def g(xs: LinkedList):
    while xs is not None:
        x, rest = xs
        yield x
```

Fixed version:
```py
def g(xs: LinkedList):
    while xs is not None:
        x, rest = xs
        yield x
        xs = rest
```
When the last line is added, this is the error I get:
![image](https://user-images.githubusercontent.com/42166884/102716481-3a6f7f00-42ed-11eb-9c8b-910193719ef0.png)

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by decorator-factory** [comment]
Can you tell why that happens?
This actually kind of makes sense to not be right, because in the `while` block, `xs`  is not `None`, but I'm potentially assigning it to `None`. On the other hand, it could be marked as possibly `None` just to the rest of the block?

**Comment by erictraut** [comment]
From a type consistency perspective, your code is fine. You are correct that you are potentially assigning `None` to `xs` within the while loop, but the condition at the top of the while loop breaks out of the loop if `xs` contains `None`. This allows the type checker to narrow the type of `xs` to `tuple[int, LinkedList]` by the time it's used in the statement `x, rest = xs`.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.96, which I just published. It will also be in the next published release of Pylance.

**Comment by decorator-factory** [comment]
Thanks!
